tests/tcl/clash_namespace.tcl

```diff
@@ -0,0 +1,31 @@
+namespace eval clash {
+    # Populate a namespace with a Clash-generated Tcl interface.
+    # Namespace is clash::tclIface::$top::$baseName
+    proc loadTclIface {top clashTclFile} {
+        # Evaluate script code inside temporary throwaway namespace to
+        # separate its definitions from ours and reduce the chance of
+        # accidentally corrupting our code.
+        namespace eval tmp "source $clashTclFile"
+        set baseName [file rootname [file tail $clashTclFile]]
+        tmp::createNamespace [namespace current]::tclIface::${top}::${baseName}
+        namespace delete tmp
+    }
+
+    proc runClashScripts {} {
+        # Identical names means identical IP, only one run needed even if it
+        # occurs in multiple HDL directories.
+        set seen [list]
+        foreach topNs [namespace children tclIface] {
+            foreach tclIface [namespace children $topNs] {
+                set api [subst $${tclIface}::api]
+                set ipName [subst $${tclIface}::ipName]
+                if {$ipName in $seen} {
+                    continue
+                }
+                ${tclIface}::createIp
+                lappend seen $ipName
+            }
+        }
+        return $seen
+    }
+}

```

Comment:
This is not an implementation of the API I designed. It should be this (adapted to the current CI script, not exactly a match for the CI script Martijn and I wrote):
```suggestion
namespace eval clash {
    # Populate a namespace with a Clash-generated Tcl interface.
    # Namespace is clash::tclIface::$top::$baseName
    proc loadTclIface {top clashTclFile} {
        # Evaluate script code inside temporary throwaway namespace to
        # separate its definitions from ours and reduce the chance of
        # accidentally corrupting our code.
        namespace eval tmp "source $clashTclFile"
        set baseName [file rootname [file tail $clashTclFile]]
        tmp::createNamespace [namespace current]::tclIface::${top}::${baseName}
        namespace delete tmp
    }

    proc runClashScripts {} {
        if {![namespace exists tclIface]} {
            # There are no scripts
            return
        }

        # Identical names means identical IP, only one run needed even if it
        # occurs in multiple HDL directories.
        set seen [list]
        foreach topNs [namespace children tclIface] {
            foreach tclIface [namespace children $topNs] {
                set api [subst $${tclIface}::api]
                if {$api ne {1}} {
                    error "Error: $tclIface doesn't implement an API we\
                        support: api = \"$api\"."
                }
                set purpose [subst $${tclIface}::scriptPurpose]
                if {$purpose ne {createIp}} {
                    error "Error: $tclIface::scriptPurpose bogus value\
                        \"$purpose\"."
                }
                set ipName [subst $${tclIface}::ipName]
                if {$ipName in $seen} {
                    continue
                }
                ${tclIface}::createIp $ipName
                lappend seen $ipName
            }
        }
        return $seen
    }
}
````

(I think the diff is not more helpful because for arcane and unknowable reasons, the script in the repository has DOS line endings. This seems to trip up GitHub's diff functionality.)

Reply:
What does checking the `$purpose` variable accomplish? I got rid of it because I thought it didn't do anything useful. 

Also this forces the API version to be `1`, which doesn't really enforce much (and IP cores are already already versioned!)

Reply:
The `tclIface::...::api` has nothing to do with Xilinx IP versions. It is the version of the Clash<->Tcl API that the generated script implements.

The idea is to lay the ground work for a future-proof Clash<->Tcl API. So when we need it for something else than Vivado's `create_ip` thing, we can bump the API, introduce a new script purpose and our scripts would check API and script purpose and invoke it in the correct manner. I wanted to codify `scriptPurpose` right now so this is established already when we later want to have more options.

And similarly, if we later need changes to how `tclIface::...::createIp` works, we bump the API as well. Our integration scripts should support all versions of the API so that when a user has created their own stuff that implements the API, they are not forced to upgrade their own stuff but our scripts gracefully handle the not-latest API they generate.

Properly checking values that currently have only one allowed option is simply defensive programming to hopefully give a meaningful error message if there is a mistake a user or we made, rather than blowing up spectacularly and unfathomably later.

Reply:
Oh I see, that makes sense.